libass/ass_cache.c

```diff
@@ -493,60 +564,92 @@ void ass_cache_dec_ref(void *value)
         return;
 
     Cache *cache = item->cache;
-    if (cache) {
-        pthread_mutex_lock(&cache->mutex);
-
-        if (item->next)
-            item->next->prev = item->prev;
-        *item->prev = item->next;
-
-        cache->items--;
-        cache->cache_size -= item->size + (item->size == 1 ? 0 : CACHE_ITEM_SIZE);
-
-        pthread_mutex_unlock(&cache->mutex);
+    if (!cache) {
+        destroy_item(item->desc, item);
+        return;
     }
 
-    destroy_item(item->desc, item);
+    CacheItem *next = NULL;
+    if (!atomic_compare_exchange_strong(&item->free_next, &next, (CacheItem*)-1))
```

Comment:
This is not possible: an item can only be added to the free list while its refcount reaches 0, which it can only do while not already in the free list.